ENH: stats: quartile coeff dispersion

YarivLevy81 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes #13385

#### What does this implement/fix?
<!--Please explain your changes.-->
Added implementation for quartile coefficient of dispersion, as requested in #13385.
Added test cases, documentation like requested in contributing guidelines.
Tried to generalise things as good as possible, and to have good code coverage - It's not a very complicated issue so I hope it's good enough.

#### Additional information
<!--Any additional information you think is important.-->
I wasn't sure about few things:
1. We could possibly implement it as `quantile_coeff_dispersion` with other values expect of 0.25, 0.5, 0.75. What are your thoughts?
2. my function returns a float or an array_like of float, is it problematic? Is it better to just return an array with a single value?
3. I'm using `np.quantile` function to compute the quartiles, this function has more arguments and but I don't necessarily want to use all of them. What would be the best practice here?

mdhaber commented on scipy/stats/__init__.py:
```diff
@@ -271,6 +271,7 @@
    combine_pvalues
    jarque_bera
    page_trend_test
+   quartile_coeff_dispersion
```
Does this belong in statistical tests, or might summary statistics be more appropriate?

YarivLevy81 replied:
Yea, I guess summary statistics is more appropriate.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,74 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+        The default is to compute with ￿Q2 = 0.75 (upper quartile), Q1 = 0.25 (lower
+        quartile).
+    axis : {int, tuple of int, None}, optional
+        Axis or axes along which the quantiles are computed. The
+        default is to compute the quartile(s) along a flattened
+        version of the array.
```
Although it is not consistent, I think the preferred default these days is `axis=0`. You can explain here that `None` is available and that it acts over a flattened version of the array, though.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,74 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
     return ((high - low)/2.0/pi) * sqrt(-2*log(R))
+
+
+def quartile_coeff_dispersion(samples, q=(0.25, 0.75), axis=None, interpolation='lower'):
```
Please use `flake8` to check for PEP8 issues. These will cause a CI failure:

![image](https://user-images.githubusercontent.com/6570539/106397793-11e6a480-63c4-11eb-9e79-ae72cb770c6b.png)

This and a few other lines are too long. You only need to fix the ones that correspond with your changes, though. (We weren't as careful in the past about this sort of thing.)

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,74 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+
+    Given two quartiles Q1, Q2 from array samples such that Q1 < Q2, the quartile
+    coefficient of dispersion is calculated by (Q2 - Q1)/(Q1 + Q2).
+    by default it is computed using 3rd and 1st (Q2, Q1 respectively) quartiles.
```
```suggestion
    By default it is computed using 3rd and 1st (Q2, Q1 respectively) quartiles.
```
Consider whether the order here (Q2, Q1) should match the order above; e,g, "By default, it is computed using the 1st and 3rd..."
Also, Q1, Q2, and Q3 are common abbreviations for 25th percentile, median, and 75th percentile, so it may be confusing to refer to the 75th percentile as Q2 here.

YarivLevy81 replied:
Good point.
Maybe using `Q_high`, `Q_low` would be better.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,74 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+        Input array.
+    q : array_like of float, optional
+        Sequence of 2 quartiles, one of [0.25, 0.5, 0.75].
+        The default is to compute with ￿Q2 = 0.75 (upper quartile), Q1 = 0.25 (lower
```
```suggestion
        The default is to compute with Q2 = 0.75 (upper quartile), Q1 = 0.25 (lower
```

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,74 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    Returns
+    -------
+    dispersion : {float, array_like of float}
+        Dispersion for each axis.
```
```suggestion
        Dispersion for each axis slice.
```
I'm not sure that "for each axis" is strictly accurate (e.g. there are many more elements in the array than there are axes). There is some precedent for using "axis slice" for what this is saying, though.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,74 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    0.15999999999999998
+
+    """
+    if len(q) != 2:
```
This might be fine, but what does happen if the user provides a `2 x n` array? If the error message is confusing, consider checking for `shape`.

YarivLevy81 replied:
Missed that one, I'll check the shape here.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,74 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    quartiles = np.quantile(samples, q=(min(q), max(q)),
+                            axis=axis, interpolation=interpolation)
+
+    def _calculate_dispersion(quartile_low, quartile_high):
+        return (quartile_high - quartile_low)/(quartile_high + quartile_low)
+
+    return _calculate_dispersion(quartiles[0], quartiles[1])
```
```suggestion
    return (quartile_high - quartile_low)/(quartile_high + quartile_low)
```
Unless there is reason not to?

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,74 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    if not np.all(np.isin(q, (0.25, 0.5, 0.75))):
+        raise ValueError("q values must be 2 from (0.25, 0.5, 0.75)")
+
+    if q[0] == q[1]:
```
Is it intentional to allow the user to provide the quantiles in either order?

YarivLevy81 replied:
It was intentional. 
What do you think would be the best practice here?

mdhaber replied:
That's a really tough one. Is there a chance that the user would expect it to return a negative number if q[1] < q[0]? Is there a reasonable interpretation of a negative value? 

Usually the safest way to go is to make it more restrictive: specify in the documentation that q[0] must be less than q[1] (to avoid the appearance of ambiguity), and writing the code to enforce that (because users expect the code to agree with the documentation). This risks the possibility that the code is a little less convenient in some cases, but it may save users from accidentally making mistakes (in cases we might not be able to imagine).

YarivLevy81 replied:
It's really tricky, I don't really see the value of a negative value here. 
Let's go with specification in the documentation + enforcing.

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1987,6 +1987,79 @@ def test_circfuncs_unit8(self):
+        # Q1 (0.25) = 4, Q2 (0.5) = 8 -> dispersion = (8-4)/(8+4) = 4/12 = 1/3
+        assert_equal(stats.quartile_coeff_dispersion(arr, q=(0.25, 0.5)), 1/3)
+
+        ref = [2, 4, 6, 8, 10, 12, 14, 16]
+        arr = np.array([ref] * len(ref)).transpose()
+
+        # Same as above but with arr as repeated columns, expecting len(arr) 1/3 dispersion values
+        assert_array_almost_equal(stats.quartile_coeff_dispersion(arr, q=(0.25, 0.5), axis=0), [1/3] * len(ref))
+
+        arr = [1.6, 2.1, 2.3, 2.4, 2.6, 2.9, 2.98, 3]
+
+        # Q1 (0.25) = 2.1, Q2 (0.5) = 2.4 -> dispersion = (2.4-2.1)/(2.4+2.1) = 0.3/4.5 ~= 0.0666
+        assert_almost_equal(stats.quartile_coeff_dispersion(arr, q=(0.25, 0.5)), (2.4 - 2.1)/(2.4 + 2.1))
```
It's OK to leave these in, but I don't know that the behavior on multidimensional arrays needs to be checked for all values of the `qs`, as there's no reason to suspect that changing these would break the way the functions works on arrays.

YarivLevy81 replied:
That's right, should I remove? It may add some granularity to the UT's - maybe better to change to different dispersion targets?

mdhaber replied:
I'm not sure. I'd just suggest that you think about the purpose of each test. 

For example, you have already tested above that
```python3
assert_equal(stats.quartile_coeff_dispersion(arr, q=(0.25, 0.5)), 1/3)
```
and in `test_simple_many`:
```python3
assert_array_equal(stats.quartile_coeff_dispersion(arr, axis=0), [0.5] * len(ref))
```
Based on the code, I don't think there is any reason to suspect that specifying `q`  is going to break the behavior on arrays, but perhaps it doesn't hurt to check the combination of array input and custom `q`s:
```python3
assert_array_almost_equal(stats.quartile_coeff_dispersion(arr, q=(0.25, 0.5), axis=0), [1/3] * len(ref))
```
But if you've done that once, then I'm not sure that
```python3
assert_array_almost_equal(stats.quartile_coeff_dispersion(arr, q=(0.5, 0.75), axis=0), [1/5] * len(ref))
```
tests anything new, unless we suspect that NumPy might make arithmetical mistakes.

Some redundancy is OK - it's better than forgetting something. But if you're going to be testing the same sorts of things for different combinations of input array values, input array sizes, custom quantiles, and `axis` arguments, then I would prefer that you [@parametrize](https://docs.pytest.org/en/stable/parametrize.html) the tests instead of copy-pasting. That would make the test code shorter, which makes it and easier to confirm that it covers the function adequately. 

Also, I appreciate your comments stating where values come from, e.g.
```
# Q1 (0.5) = 8, Q2 (0.75) = 12 -> dispersion = (12-8)/(12+8) = 4/20 = 1/5
```
but one thing that would help is comments indicating the purpose of each part of the test. For example, is the purpose of the test exactly the same as previous tests but with different numbers, or are you testing a different feature of the code?

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1987,6 +1987,79 @@ def test_circfuncs_unit8(self):
+        # Q1 (0.5) = 2.4, Q2 (0.75) = 2.9 -> dispersion = (2.9-2.4)/(2.9+2.4) ~= 0.0943
+        assert_almost_equal(stats.quartile_coeff_dispersion(arr, q=(0.5, 0.75)), (2.9 - 2.4)/(2.9 + 2.4))
+
+    def test_bad_q_value(self):
+        arr = [2, 4, 6, 8, 10, 12, 14, 16]
+
+        # Quartiles are any of (0.25, 0.5, 0.75)
+        assert_raises(ValueError, stats.quartile_coeff_dispersion, arr, q=(0.3, 0.5))
+
+        # Quartile values must differ (otherwise, it's constant zero)
+        assert_raises(ValueError, stats.quartile_coeff_dispersion, arr, q=(0.5, 0.5))
+
+    def test_bad_q_len(self):
+        arr = [2, 4, 6, 8, 10, 12, 14, 16]
+
+        # Expecting 2 quartiles
+        assert_raises(ValueError, stats.quartile_coeff_dispersion, arr, q=(0.5, ))
+
+
```
I'd suggest a single input validation function. Please use `with assert_raises(match="This is the error message I'm looking for"):` as a context manager rather than as a function for more specific checks.

YarivLevy81 replied:
Done that, should I check for more details of the raised exceptions?

mdhaber commented:
> We could possibly implement it as quantile_coeff_dispersion with other values expect of 0.25, 0.5, 0.75. What are your thoughts?

I'd prefer to leave it unrestricted, personally. The restriction seems unnecessary and it's simpler in documentation and code to leave out the restriction. If there is a strong reason to protect people from using the 20th and 80th percentiles, please correct me!

> my function returns a float or an array_like of float, is it problematic? Is it better to just return an array with a single value?

It is preferred to return a float when the axis of a 1d array gets consumed. That is the typical behavior of `np.mean([1, 2, 3])`, for example, unless `keepdims=True`, and I don't think we have functions with `keepdims=True` like NumPy does.

> I'm using np.quantile function to compute the quartiles, this function has more arguments and but I don't necessarily want to use all of them. What would be the best practice here?

Do you mean "should I expose them as parameters of this function?" It's OK to leave them out. For the sake of keeping the API simpler, I think it's preferable to leave them out.

mdhaber commented:
Nice start @YarivLevy81 ! Don't worry about some of the CI failures, as we're having issue in master right now due to the NumPy 1.20 release, but due check out the "scipy.scipy (Lint)" failure (PEP8).

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,73 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+        Input array.
+    q : array_like of float, optional
+        Sequence of 2 quartiles,
+        The default is to compute with ￿Q3 = 0.75, Q1 = 0.25.
```
It looks like you tried to fix some of these things, but maybe the changes were partially reversed in a subsequent commit? I'm still seeing the special character here:
![image](https://user-images.githubusercontent.com/6570539/106503781-f25e8300-647a-11eb-8b8e-d5451a7ad26e.png)

 and Q3/Q1 instead of Q_high/Q_low.  Please take a look at the previous comments and check that they were addressed as you intended. Let me know when you'd like me to take another look!

YarivLevy81 replied:
@mdhaber, I tried to address every issue. still struggling with my failing CI tests regarding the refguide.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,72 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    if np.shape(q) != (2,):
+        raise ValueError("Shape of q must be (,2)")
+
+    if (q[0] > 1 or q[1] > 1) or (q[0] < 0 or q[1] < 0):
```
```suggestion
    if q[0] < 0 or q[1] > 1:
```
I think the next condition should take care of the rest, right?

YarivLevy81 replied:
Totally right.

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1987,6 +1987,98 @@ def test_circfuncs_unit8(self):
+        arr = [1.6, 2.1, 2.3, 2.4, 2.6, 2.9, 2.98, 3]
+
+        # Q1 = 2.1, Q2 = 2.9 -> dispersion = (2.9-2.1)/(2.9+2.1) = 0.8/5 ~= 0.16
+        assert_almost_equal(stats.quartile_coeff_dispersion(arr), 0.16)
```
The documentation for `assert_almost_equal` states:
![image](https://user-images.githubusercontent.com/6570539/106830188-6b5c0700-6642-11eb-8ad3-a6b979d7fb70.png)

Old code uses these old functions, but I'd suggest `assert_allclose` if it does what you want it to.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,72 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Quartile_coefficient_of_dispersion
```
I try to recommend slightly more formal references. For instance, the Wikipedia citation in geometric mean looks like:
![image](https://user-images.githubusercontent.com/6570539/106831240-6ac47000-6644-11eb-86ca-d583d084c461.png)

mdhaber reviewed (changes requested):
A few other things to think about, but I think after this it will be very close!

YarivLevy81 commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,72 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    >>> quartile_coeff_dispersion(samples=[2, 4, 6, 8, 10, 12, 14, 16])
+    0.5
+
+    >>> quartile_coeff_dispersion(samples=[1.6, 2.1, 2.3, 2.4,
+    ...                           2.6, 2.9, 2.98, 3])
+    0.15999999999999998
+
```
That's a good point, I'm changing it.

YarivLevy81 commented:
@mdhaber, thanks again for the guidance and the great points.
I've tried to address all issues, mainly I hope the my tests now make more sense,  the `other_interpolation` test might be redundant. Anyways, I renamed tests and commented the purpose of each test, removed some redundancy to make some sense.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+
+    References
+    ----------
+    .. [1] "Quartile Coefficient of Dispersion", *Wikipedia*,
+    https://en.wikipedia.org/wiki/Quartile_coefficient_of_dispersion.
```
```suggestion
    .. [1] "Quartile Coefficient of Dispersion", *Wikipedia*,
           https://en.wikipedia.org/wiki/Quartile_coefficient_of_dispersion.
```

mdhaber replied:
To fix this, just click "Add suggestion to batch" for this and any suggestions below that you agree with, then "Commit suggestions"

mdhaber commented:
The Linux_Python_37_32bit_full failure is not this PR; it's an intermittent failure I'm addressing in gh-13526.

I'm not sure about the ci/circleci: build_docs failure. It might be related to the refguide failure (see inline comment), it might be a temporary issue with CircleCI, or it might be something that will be harder to debug. Let's see if it goes away on its own.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3526,24 +3528,26 @@ def quartile_coeff_dispersion(samples, q=(0.25, 0.75), axis=0,
-    >>> quartile_coeff_dispersion(samples=[1.6, 2.1, 2.3, 2.4,
-    ...                           2.6, 2.9, 2.98, 3])
-    0.15999999999999998
+    >>> quartile_coeff_dispersion(a=[1.6, 2.1, 2.3, 2.4,
+    ...                           2.6, 2.9, 2.98, 3],
```
```suggestion
    >>> quartile_coeff_dispersion(a=[1.6, 2.1, 2.3, 2.4,
    ...                              2.6, 2.9, 2.98, 3],
```

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    a : array_like
+        Input array.
+    q : array_like of float, optional
+        Sequence of 2 quantiles, such that q[0] < q[1].
+        The default is to compute with Q_high = 0.75, Q_low = 0.25.
```
```suggestion
        Sequence of 2 quantiles, (Q_low, Q_high), such that Q_low < Q_high.
        The default is to compute with Q_low = 0.25, Q_high = 0.75
```
First line suggestion is because relationship between `q` and (Q_low, Q_high) was not explicitly established here. I think this is just slightly more clear. 
Second is because Q_low appears first everywhere else (except as necessary in the formula). 100% personal nitpick; so you're welcome to disagree if you prefer the other way.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+        The default is to compute with Q_high = 0.75, Q_low = 0.25.
+    axis : {int, tuple of int, None}, optional
+        Axis along which the coefficient of dispersion is computed.
+        Default is 0. If None, compute over the whole array a.
```
```suggestion
        Default is 0. If None, compute over the whole array `a`.
```
Referenced parameters [get single backticks](https://numpydoc.readthedocs.io/en/latest/format.html). (Code gets double backticks.)

I think "Q_low" and "Q_high`" are fine as they are because they are neither really code nor parameters.

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1987,6 +1987,88 @@ def test_circfuncs_unit8(self):
+        # Quartiles are bounded by 0 < q < 1
+        with assert_raises(ValueError) as err:
+            stats.quartile_coeff_dispersion(a=arr, q=(0.2, 1.5))
+        err.match("q values must be 2 floats in range \\(0.0, 1.0\\)")
```
I've actually not used the `match` method like this before (I've done it as an argument of `pytest.raises`. Does it work the same way?

YarivLevy81 replied:
I believe they work the same, I can change it to standard to avoid confusion.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    quartiles = np.quantile(a, q=(q[0], q[1]),
+                            axis=axis, interpolation=interpolation)
+
+    return (quartiles[1] - quartiles[0])/(quartiles[1] + quartiles[0])
```
Question to other reviewer: are warnings on divide by zero OK? Result is nan, as I suppose it should be.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    a : array_like
+        Input array.
+    q : array_like of float, optional
+        Sequence of 2 quantiles, such that q[0] < q[1].
```
I think the use of `quantiles` and `quartiles` is appropriate, btw. It's unfortunate that the words are so similar, but I think the distinction is made appropriately here.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    Given two quartiles Q_low, Q_high from array samples such that
+    Q_low < Q_high, the quartile coefficient of dispersion is calculated by
+    (Q_high - Q_low)/(Q_high + Q_low).
+    by default it is computed using 3rd and 1st quartiles.
```
```suggestion
    By default, it is computed using 3rd and 1st quartiles.
```

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    """
+    Compute the quartile coefficient of dispersion for a given sample array.
+
+    Given two quartiles Q_low, Q_high from array samples such that
```
For the record, I think the rendering of "Q_low" and "Q_high" as regular text are fine because they are neither code nor parameters. Personally, when reading documentation online, I'd prefer to see them as mathematical symbols (e.g. `:math: Q_{\text{low}}`), but please do not change to that because it would be difficult to read for those who use documentation in the console / IDE. I suppose they could be rendered as code, but I don't think that is necessary.

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1987,6 +1987,88 @@ def test_circfuncs_unit8(self):
+
+        # Interpolation = 'midpont', Q_low (0.25) = (3+2)/2 = 2.5,
+        # Q_high (0.75) = (7+6)/2 = 6.5 => dispersion = 4/9
+        assert_equal(stats.quartile_coeff_dispersion(arr, interpolation='midpoint'), 4/9)
+
```
```suggestion
        assert_equal(stats.quartile_coeff_dispersion(arr, interpolation='midpoint'), 4/9)
        assert_equal(stats.quartile_coeff_dispersion(arr), 3.5/9)

```

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    0.0638297872340425
+
+    """
+    if np.shape(q) != (2,):
+        raise ValueError("Shape of q must be (2,)")
+
+    if q[0] < 0 or q[1] > 1:
+        raise ValueError("q values must be 2 floats in range (0.0, 1.0)")
+
+    if q[0] >= q[1]:
+        raise ValueError("q values should satisfy q[0] < q[1]")
```
I think this is sufficient input validation because NumPy will give reasonable error messages for invalid values of the other parameters.

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1987,6 +1987,88 @@ def test_circfuncs_unit8(self):
         assert_allclose(stats.circstd(x, high=180), 20.91551378, rtol=1e-7)
 
 
+class TestQuartileCoeffOfDispersion(object):
```
Tests are carefully explained; thanks!

I think they're sufficient. For instance, I don't think there's a need to test exotic nd-inputs because you've already shown correct behavior for 2d and the code is straightforward enough to see that it should work as long as `np.quantile` does what it says it does:
> If multiple quantiles are given, first axis of the result corresponds to the quantiles. The other axes are the axes that remain after the reduction of _a._

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    Returns
+    -------
+    dispersion : {float, array_like of float}
+        Dispersion for each axis slice.
```
I think this is consistent with the level of detail given in other summary statistics functions. NumPy is much more explicit about the resulting shape, but if we want that sort of thing, it can be done at the same time as all the other summary statistics functions.

mdhaber reviewed (approved):
Looks good to me after the suggestions are implemented (assuming CI passes).

YarivLevy81 commented:
> Looks good to me after the suggestions are implemented (assuming CI passes).
All checks pass.

mdhaber commented on scipy/stats/tests/test_morestats.py:
```diff
@@ -1987,6 +1987,91 @@ def test_circfuncs_unit8(self):
+
+        arr = [1, 2, 3, 4, 5, 6, 7, 8]
+
+        # Interpolation = 'midpont', Q_low (0.25) = (3+2)/2 = 2.5,
+        # Q_high (0.75) = (7+6)/2 = 6.5 => dispersion = 4/9
+        assert_equal(stats.quartile_coeff_dispersion(arr, interpolation='midpoint'), 4/9)
+        # Interpolation = 'midpont', Q_low (0.25) = 2
+        # Q_high (0.75) = 6 => dispersion = 4/8
+        assert_equal(stats.quartile_coeff_dispersion(arr), 4/8)
```
```suggestion
        # Interpolation = 'midpoint', Q_low (0.25) = (3+2)/2 = 2.5,
        # Q_high (0.75) = (7+6)/2 = 6.5 => dispersion = 4/9
        assert_equal(stats.quartile_coeff_dispersion(arr, interpolation='midpoint'), 4/9)
        # Interpolation = 'lower' (default), Q_low (0.25) = 2
        # Q_high (0.75) = 6 => dispersion = 4/8
        assert_equal(stats.quartile_coeff_dispersion(arr), 4/8)
```
Oops, I didn't realize that lower was the default when I wrote the suggestion. Thanks for correcting it.

mdhaber replied:
Sorry if I'm forgetting; we might have discussed it before - curious why that is the default?

YarivLevy81 replied:
Hmm, I tried to find what might be the best default interpolation.
I took a look on the [wiki](https://en.wikipedia.org/wiki/Quartile_coefficient_of_dispersion) page, and the examples (which were pretty simple) seemed to use `lower` interpolation - therefore I followed.

This is a rather tricky one, the default interpolation value for `numpy.quantile` is actually `linear`. I'd be happy to change it if you suggest we comply with numpy.

mdhaber replied:
Interesting... actually those Wikipedia examples don't seem to use any of the available interpolation options, right? It's possible the author intended for there to be more data points, e.g.
```
a = [0, 2, 4, 6, 8, 10, 12, 14, 16]
```
Then no interpolation would be needed because the 25th and 75th percentiles lie right on the data. Do you think they made a mistake?

mdhaber replied:
Hmm... actually, the definition of the 1st and 3rd quartiles I'm finding in some places is not identical to 25th and 75th percentiles.

YarivLevy81 replied:
Interesting, I wasn't familiar with this one. 
I think I kind of implemented Method 4 from the wiki page, didn't I?

YarivLevy81 replied:
Do we want to implement all different methods? I could do it.

mdhaber replied:
Using no.quantile you get method 4, yes.
I don't think you need all four, no. I will make a suggesting to the documentation later today, and i think this will be resolved.

mdhaber commented:
@rkern I think this PR is in pretty good shape. Would you like to take a look?

mdhaber commented:
Actually, one thought after looking at the reference linked from Wikipedia - it might be desirable in the future to provide a way of computing a coefficient of dispersion confidence interval. If that's the case, please think about how that might be possible in a backwards-compatible way. One option is to have an optional argument that selects the option to return the confidence interval and, if that's the case, the function returns the confidence interval rather than just the sample statistic.

mdhaber commented on scipy/stats/morestats.py:
```diff
@@ -3478,3 +3478,76 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
+    Given two quartiles Q_low, Q_high from array samples such that
+    Q_low < Q_high, the quartile coefficient of dispersion is calculated by
+    (Q_high - Q_low)/(Q_high + Q_low).
+    By default, it is computed using 3rd and 1st quartiles.
```
```suggestion
    By default, it is computed using the 1st and 3rd quartiles, that is,
    the 0.25 and 0.75 _quantiles_ as calculated by `np.quantile`. 
```

mdhaber commented:
@smurfit89 Does this address your proposal on gh-13385?

mdhaber reviewed (commented):
I've been taking another look at this today, and I think we need to address the discrepancy in quartile calculation conventions. Currently this function cannot replicate the result given in [1]. I submitted YarivLevy81/scipy#1 to begin to address this, but I think we need to take a closer look at the literature and see if there is consensus on how to define the quartiles.

mdhaber reviewed (changes requested):
Please see YarivLevy81/scipy#1

mdhaber commented:
@YarivLevy81 were you still interested in completing this? If not, I understand.
